### PR TITLE
Converted to a Ruby Gem.

### PR DESCRIPTION
**Pull request by panthomakos**
This converts the aws-codedeploy-agent into an actual RubyGem that can be installed w/o a need for RPM/DEB packages and is compatible with multiple Ruby versions instead of just Ruby 2.0. I've tested this with the actual AWS CodeDeploy service.  Sorry the PR is so large - I removed a lot of the vendor files since they were not necessary.  I believe there are still many simplifications available, but this should be a good first step at making this a usable gem.

You can test this exact PR by following the README instructions.

https://rubygems.org/gems/aws-codedeploy-agent
- Updated the Gemfile so that it uses the gemspec as per Gem best
  practices.
- Overhauled the gemspec file so that it actually works and properly
  includes the source and binary files.
- Made gem compatible with all versions of Ruby. The ruby-2.0 package
  was not a real dependency of this code and not having that explicitly
  built into the executables makes the gem more useful as it can run
  on a variety of Ruby versions.
- Removed the auto-update feature of the `init.d` script and generally
  simplified the script. Updates should probably happen explicitly by
  running `gem install` instead.
- Added executables to install/uninstall as a SYSVINIT service.

### CODE REVIEW

File: bin/codedeploy-agent [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/ruby2.0
+#!/usr/bin/env ruby
```
Comment (panthomakos) [inline bin/codedeploy-agent]: This is the recommended way to use the user/system defined ruby.

**Comment by suryanarayanan** [comment]
Hi,
Thanks for contributing to aws-codedeploy-agent. We are in the process of reviewing it internally. Meanwhile just wanted to make sure if you are okay with publishing this under Apache 2.0 license. Thanks again.

**Comment by panthomakos** [comment]
@suryanarayanan 

Thanks for the reply.  I am okay w/ publishing under the Apache 2.0 license.

**Comment by jdrago999** [comment]
I just tried this out on an Ubuntu 14.04 instance which was working fine with the *.deb version of the codedeploy agent.

Installation was fine, but I see this in /var/log/aws/codedeploy-agent/codedeploy-agent.log

2015-02-23 04:04:43 ERROR [codedeploy-agent(4705)]: InstanceAgent::CodeDeployPlugin::CommandPoller: Cannot reach InstanceService: Aws::CodeDeployCommand::Errors::AccessDeniedException - Aws::CodeDeployCommand::Errors::AccessDeniedException
2015-02-23 04:04:45 INFO  [codedeploy-agent(4705)]: [Aws::CodeDeployCommand::Client 400 0.014956 0 retries] poll_host_command(host_identifier:"arn:aws:ec2:us-east-1:xxxxxx:instance/i-xxxxxxxx") Aws::CodeDeployCommand::Errors::AccessDeniedException Aws::CodeDeployCommand::Errors::AccessDeniedException


ubuntu@ip-10-0-5-51:~/aws-codedeploy-agent$ ps -ef | grep code
root      4697     1  0 04:04 pts/1    00:00:00 codedeploy-agent: master 4697
root      4705  4697  0 04:04 pts/1    00:00:00 codedeploy-agent: InstanceAgent::CodeDeployPlugin::CommandPoller of master 4697

**Comment by panthomakos** [comment]
@jdrago999 Those (`Aws::CodeDeployCommand::Errors::AccessDeiedException`) sound like permission issues to me. You probably don't have the correct IAM role configured.

I am happy to revise my PR to adjust permissions or instructions if you can provide me with more details. I am able to get this working as a gem on a stock Ubuntu AMI in EC2, but I do have to run everything as root. I believe this is to be expected? If not, I am happy to figure out what I need to adjust.

```
sudo gem install aws-codedeploy-agent
sudo codedeploy-install
```

Copying to `/var/www` sounds like a directory permission error. Maybe you need to verify that the process is running as root and that you have access to create/modify those directories?

**Comment by suryanarayanan** [comment]
@panthomakos  We've got a few other things to wrap up before testing this. It is going to take some time for us to merge the PR. Sorry about the delay. I'll keep you posted. Thanks.

**Comment by panthomakos** [comment]
@suryanarayanan - no problem. Thanks for the update!

**Comment by panthomakos** [comment]
@suryanarayanan Any update on this PR?

**Comment by benjaminos** [comment]
Any chance this could be updated to support eu-west-1 region, now that CodeDeploy has gone live in Europe? I've manually hacked around it for the time being - but my Ruby is virtually non-existent, so I'm not confident in my changes :)

**Comment by panthomakos** [comment]
@benjaminos I don't know if your change falls into the scope of this PR. I'm just trying to get this project into a usable gem format. I would be happy to review your code, but I've been waiting for months to get this PR added to the code-base...

**Comment by panthomakos** [comment]
@suryanarayanan Are you still reviewing this PR? Any estimates on when the review will be complete?

**Comment by suryanarayanan** [comment]
@panthomakos We have been focused on getting the on-premises support out and expanding to Ireland and Sydney. Unfortunately we couldn't prioritize this work. I currently don't have an estimate to give you. However I'll bring this up to the team's notice. Thanks.

**Comment by jdrago999** [comment]
Looking forward to this very much.

**Comment by autrilla** [comment]
@panthomakos I'm trying to use this on CentOS 7 (which uses systemd) but when I run it, even with /usr/local/bin/codedeploy-agent start, I get an error saying "Not able to find an endpoint. Unknown region." I tried setting AWS_REGION=eu-west-1 before launching, but that doesn't change the outcome.

On a general topic, I'd really like Amazon to expand their supported distributions, considering RHEL/CentOS are way more popular than Amazon Linux.

**Comment by benjaminos** [comment]
@autrilla That's not a CentOS 7 issue, that's due to the issue I've noted above - it hasn't been updated since CodeDeploy went live in Europe, so the new regions don't exist in the gem :/

**Comment by stphung** [comment]
+1 on this, looking forward to it.

**Comment by panthomakos** [comment]
I've included the most recent changes from upstream and resolved an issue with stuck deployments. It does not seem like this PR has any interest from the AWS side, so I'll continue to keep my branch and the gem up to date and hopefully in working order.

**Comment by nwalke** [comment]
@suryanarayanan Any updates here?  Sticking this to Ruby 2.0 makes it so that we have to install multiple versions of Ruby which isn't super ideal.

**Comment by panthomakos** [comment]
My branch has been rebased on master.

**Comment by bdashrad** [comment]
👍  This would be great to have. I hate needing to install multiple versions of ruby when I can avoid it.

**Comment by davividal** [comment]
Any update on this?

**Comment by nwalke** [comment]
@suryanarayanan @feverLu Would it be possible to get this merged now that it's had two years of soak?  Good grief...

**Comment by Joeasaurus** [comment]
I agree with @nwalke, the current install method gives me great grief on CentOS 6 where I need to install Puppet (depending on OS-ruby 1.8.7) and a separate ruby 2.x. Both want to put a binary at /usr/bin/ruby.

It'd be nice to choose where I put the ruby it depends on.

**Comment by panthomakos** [comment]
I will no longer maintain my fork of this project. AWS does not appear interested in releasing as a RubyGem.

**Comment by nwalke** [comment]
@asaf-erlich It's a shame to see such a great effort let go like this.  Can we really not get any input from the AWS side?  This is your product that a member of the community is trying to improve, for free.

I don't get it.

**Pull request closed without merging**